// ceri/lib/c-for.js
var isFunction;

isFunction = require("./_helpers").isFunction;

module.exports = {
  _name: "c-for",
  _v: 1,
  mixins: [require("./structure"), require("./for")],
  _elLookup: {
    cFor: {
      extract: {
        "": ["template", "iterate", "names", "computed", "id", "tap"]
      },
      cb: function(o, arg) {
        var children, comment, template;
        children = arg.children;
        comment = document.createComment("c-for");
        if (isFunction(children)) {
          template = children;
        } else {
          template = o.template;
        }
        this.$nextTick(function() {
          var ref, scopes, tap;
          scopes = this.$for({
            anchor: comment,
            template: template,
            value: o.iterate,
            names: (ref = o.names) != null ? ref.split(",") : void 0,
            computed: o.computed,
            id: o.id
          }).scopes;
          if ((tap = o.tap) != null) {
            return this.$path.setValue({
              path: tap,
              value: scopes
            });
          }
        });
        return {
          el: comment,
          options: null
        };
      }
    }
  }
};

;
